src/librustc/middle/const_eval.rs

```diff
@@ -607,6 +697,71 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
     Ok(result)
 }
 
+fn resolve_trait_associated_const<'a, 'tcx: 'a>(tcx: &'a ty::ctxt<'tcx>,
+                                                ti: &'tcx ast::TraitItem,
+                                                trait_id: ast::DefId,
+                                                ref_id: ast::NodeId)
+                                                -> Option<&'tcx Expr>
+{
+    let rcvr_substs = ty::node_id_item_substs(tcx, ref_id).substs;
```

Comment:
So you're expecting substitutions to be there, which means this can't be used earlier...
But this is an `ExprPath`, isn't it? Which means that *at least* the `<T as Trait>::CONST` case could be handled here, if an `astconv` interface were provided.

That reminds me, I meant to check if it's possible to move `const_eval` and everything depending on it to `rustc_typeck` and integrate it properly.
Maybe this can all be done in a follow-up PR after the groundwork has been laid.